pkg/registration/hub/gc/gc_resources.go

```diff
@@ -0,0 +1,228 @@
+package gc
+
+import (
+	"context"
+	stderrors "errors"
+	"fmt"
+	"sort"
+	"strconv"
+
+	"github.com/openshift/library-go/pkg/operator/events"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/metadata"
+	"k8s.io/client-go/util/retry"
+
+	clusterv1 "open-cluster-management.io/api/cluster/v1"
+)
+
+type gcResourcesController struct {
+	metadataClient metadata.Interface
+	resourceList   []schema.GroupVersionResource
+	eventRecorder  events.Recorder
+}
+
+func newGCResourcesController(
+	metadataClient metadata.Interface,
+	resourceList []schema.GroupVersionResource,
+	eventRecorder events.Recorder,
+) gcReconciler {
+	return &gcResourcesController{
+		metadataClient: metadataClient,
+		resourceList:   resourceList,
+		eventRecorder:  eventRecorder.WithComponentSuffix("gc-resources"),
+	}
+}
+func (r *gcResourcesController) reconcile(ctx context.Context, cluster *clusterv1.ManagedCluster) (gcReconcileOp, error) {
+	if cluster.DeletionTimestamp.IsZero() {
+		return gcReconcileContinue, nil
+	}
+
+	for _, resource := range r.resourceList {
+		err := r.deleteResources(ctx, resource, cluster.Name)
+		if err != nil {
+			return gcReconcileContinue, fmt.Errorf("failed to delete %v in %s ns, %v",
+				resource.Resource, cluster.Name, err)
+		}
+
+		remainingCnt, finalizerPendingCnt, err := r.RemainingCnt(ctx, resource, cluster.Name)
+		if err != nil {
+			return gcReconcileContinue, fmt.Errorf("failed to count %v in %s ns, %v",
+				resource.Resource, cluster.Name, err)
+		}
+
+		if remainingCnt == 0 {
+			continue
+		}
+
+		meta.SetStatusCondition(&cluster.Status.Conditions, metav1.Condition{
+			Type:   ManagedClusterConditionDeleting,
+			Status: metav1.ConditionFalse,
+			Reason: "ResourceRemaining",
+			Message: fmt.Sprintf("The resource %v is remaning, the remaining count is %v, "+
+				"the finalizer pending count is %v", resource.Resource, remainingCnt, finalizerPendingCnt),
+		})
+
+		return gcReconcileRequeue, nil
+	}
+
+	return gcReconcileContinue, nil
+}
+
+func (r *gcResourcesController) deleteResources(ctx context.Context,
+	resource schema.GroupVersionResource, clusterName string) error {
+	labelSelectors, err := r.generateLabelSelectors(ctx, resource, clusterName)
+	if err != nil {
+		return err
+	}
+
+	for _, labelSelector := range labelSelectors {
+		err, deleted := r.deleteResourcesWithLabel(ctx, resource, clusterName, labelSelector)
+		if err != nil {
+			return err
+		}
+		if !deleted {
+			return nil
+		}
+	}
+	return nil
+}
+
+func (r *gcResourcesController) generateLabelSelectors(ctx context.Context,
+	resource schema.GroupVersionResource, clusterName string) ([]string, error) {
+	allResources, err := r.metadataClient.Resource(resource).Namespace(clusterName).List(ctx, metav1.ListOptions{})
```

Comment:
if we also need a list here, should we list all at first so we do not need list again when calculating remaining.

Reply:
changed to list only once in each reconcile, calculate remanings, sort the resources by priority, then delete the resources with the lowest priority, and requeue.